File: judo-sdk/src/main/java/com/judopay/api/PayLoadInterceptor.java
package com.judopay.api;

import android.content.Context;
import androidx.annotation.NonNull;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.judopay.arch.PayLoadUtil;
import com.judopay.model.EnhancedPaymentDetail;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;

import okhttp3.Interceptor;
import okhttp3.MediaType;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okio.Buffer;

public class PayLoadInterceptor implements Interceptor {

    private static final String ENHANCED_PAYMENT_DETAIL = "EnhancedPaymentDetail";
    private final List<String> payLoadPaths = Arrays.asList(
            "/transactions/payments",
            "/transactions/preauths",
            "/transactions/registercard",
            "/transactions/checkcard");

    private final Context context;

    PayLoadInterceptor(@NonNull final Context context) {
        this.context = context;
    }

    @Override
    public Response intercept(@NonNull final Chain chain) throws IOException {
        final Request request = chain.request();
        final String path = request.url().encodedPath();

        if (payLoadPaths.contains(path) && request.body() != null) {

            final JsonElement jsonElement = convertRequestBodyToJson(request);
            if (jsonElement != null) {
                final JsonObject json = jsonElement.getAsJsonObject();
                if (json.get(ENHANCED_PAYMENT_DETAIL) == null) {
                    final JsonObject enhancedPaymentDetail = getEnhancedPaymentDetail();
                    json.add(ENHANCED_PAYMENT_DETAIL, enhancedPaymentDetail);
                }

                return chain.proceed(request.newBuilder()
                        .post(convertJsonToRequestBody(json))
                        .build());
            }
        }

        return chain.proceed(request);
    }

    private JsonObject getEnhancedPaymentDetail() {
        final EnhancedPaymentDetail enhancedPaymentDetail = PayLoadUtil.getEnhancedPaymentDetail(context);

        final Gson gson = new Gson();
        final String paymentDetail = gson.toJson(enhancedPaymentDetail);
        final JsonElement jsonElement = new JsonParser().parse(paymentDetail);

        return jsonElement.getAsJsonObject();
    }

    private RequestBody convertJsonToRequestBody(final JsonObject json) {
        final MediaType mediaType = MediaType.parse("application/json");

        return RequestBody.create(mediaType, json.toString());
    }

    private JsonElement convertRequestBodyToJson(final Request request) {
        final Buffer buffer = new Buffer();
        try {
            final Request copy = request.newBuilder().build();
            final RequestBody requestBody = copy.body();
            if (requestBody != null) {
                requestBody.writeTo(buffer);
                final String body = buffer.readUtf8();
                final JsonParser parser = new JsonParser();
                return parser.parse(body);
            }
        } catch (IOException ignore) {
        } finally {
            buffer.close();
        }
        return null;
    }
}

File: judokit-android/src/main/java/com/judokit/android/api/interceptor/PayLoadInterceptor.kt
package com.judopay.api.interceptor

import android.content.Context
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.google.gson.JsonSyntaxException
import com.judopay.toJSONString
import java.io.IOException
import okhttp3.Interceptor
import okhttp3.MediaType
import okhttp3.Request
import okhttp3.RequestBody
import okhttp3.Response
import okio.Buffer

private const val ENHANCED_PAYMENT_DETAIL = "EnhancedPaymentDetail"
private val PAYLOAD_ENDPOINTS = arrayListOf(
    "/transactions/payments",
    "/transactions/preauths",
    "/transactions/registercard",
    "/transactions/checkcard"
)

class PayLoadInterceptor internal constructor(private val context: Context) : Interceptor {

    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {

        val request = chain.request()
        val path = request.url().encodedPath()

        if (PAYLOAD_ENDPOINTS.contains(path) && request.body() != null) {
            convertRequestBodyToJson(request)?.let {
                val json = it.asJsonObject
                if (json[ENHANCED_PAYMENT_DETAIL] == null) {
                    json.add(ENHANCED_PAYMENT_DETAIL, enhancedPaymentDetail)
                }

                return chain.proceed(
                    request.newBuilder()
                        .post(convertJsonToRequestBody(json))
                        .build()
                )
            }
        }

        return chain.proceed(request)
    }

    private val payloadService = PayloadService(context)

    private val enhancedPaymentDetail: JsonObject
        get() {
            val paymentDetail = payloadService.getEnhancedPaymentDetail()?.toJSONString()
            return try {
                val jsonElement = JsonParser().parse(paymentDetail)
                jsonElement.asJsonObject
            } catch (e: JsonSyntaxException) {
                JsonObject()
            }
        }

    private fun convertJsonToRequestBody(json: JsonObject): RequestBody {
        val mediaType = MediaType.parse("application/json")
        return RequestBody.create(mediaType, json.toString())
    }

    private fun convertRequestBodyToJson(request: Request): JsonElement? {
        val buffer = Buffer()
        try {
            request.newBuilder().build().body()?.let {
                it.writeTo(buffer)
                val body = buffer.readUtf8()
                val parser = JsonParser()
                return parser.parse(body)
            }
        } catch (ignore: IOException) {
        } finally {
            buffer.close()
        }
        return null
    }
}
